spatialos-sdk/src/worker/schema/component_update.rs

```diff
@@ -0,0 +1,159 @@
+        T::get_update(self, field)
+    }
+
+    pub fn add_field<T: SchemaField>(&mut self, field: FieldId, value: &T::RustType) {
+        T::add_update(self, field, value);
+    }
+
+    pub fn field_array<T: ArrayField>(&self, field: FieldId) -> Option<Vec<T::RustType>> {
+        if T::has_update(self, field) {
+            Some(self.fields().field_array::<T>(field))
+        } else {
+            None
+        }
+    }
+
+    pub fn add_field_array<T: ArrayField>(&mut self, field: FieldId, value: &[T::RustType]) {
+        self.fields_mut().add_field_array::<T>(field, value)
+    }
+
+    pub fn add_cleared(&mut self, field: FieldId) {
+        unsafe {
+            Schema_AddComponentUpdateClearedField(self.as_ptr(), field);
+        }
+    }
+
+    /// Returns an iterator over any cleared fields included in this update.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # // TODO: Use a pre-existing component, rather than having to define one here.
+    /// # use spatialos_sdk::worker::{component::*, schema::{self, *}};
+    /// # use std::{thread, sync::Arc};
+    /// #
+    /// # pub struct CustomComponent;
+    /// #
+    /// # impl Component for CustomComponent {
+    /// #     const ID: ComponentId = 7777;
+    /// #     type Update = CustomComponentUpdate;
+    /// # }
+    /// #
+    /// # impl SchemaObjectType for CustomComponent {
+    /// #     fn from_object(object: &Object) -> Self { Self }
+    /// #     fn into_object(&self, object: &mut Object) {}
+    /// # }
+    /// #
+    /// # pub struct CustomComponentUpdate;
+    /// #
+    /// # impl Update for CustomComponentUpdate {
+    /// #     type Component = CustomComponent;
+    /// #
+    /// #     fn from_update(update: &ComponentUpdate) -> Self { Self }
+    /// #     fn into_update(&self, update: &mut ComponentUpdate) {}
+    /// # }
+    /// # let mut update = ComponentUpdate::new(&CustomComponentUpdate);
+    /// for cleared in update.cleared() {
+    ///     // Clear the specified field.
+    /// }
+    /// ```
+    pub fn cleared(&self) -> impl Iterator<Item = FieldId> + '_ {
+        let count = unsafe { Schema_GetComponentUpdateClearedFieldCount(self.as_ptr()) };
+
+        ClearedFieldIter {
+            update: self,
+            count,
+            index: 0,
+        }
+    }
+
+    pub(crate) fn fields(&self) -> &Object {
+        unsafe { Object::from_raw(Schema_GetComponentUpdateFields(self.as_ptr())) }
+    }
+
+    pub(crate) fn fields_mut(&mut self) -> &mut Object {
+        unsafe { Object::from_raw_mut(Schema_GetComponentUpdateFields(self.as_ptr())) }
+    }
+
+    pub(crate) fn as_ptr(&self) -> *mut Schema_ComponentUpdate {
+        self as *const _ as *mut _
+    }
+}
+
+impl OwnableImpl for ComponentUpdate {
+    type Raw = Schema_ComponentUpdate;
+
+    unsafe fn destroy(inst: *mut Self::Raw) {
+        Schema_DestroyComponentUpdate(inst);
+    }
+}
+
+unsafe impl Send for ComponentUpdate {}
+
+struct ClearedFieldIter<'a> {
```

Comment:
When we upgrade to 13.8, there's a additional API  call for interacting with cleared fields which will make this a little easier (https://docs.improbable.io/reference/13.8/releases/release-notes#worker-sdk-in-c) 

For example - instead of calling `field` for each field on the object then iterating through cleared fields, the `field` could check itself for "cleared-ness" and return the appropriate object

Reply:
That's terrific! That's what I had originally envisioned doing when I started writing the API, and had to switch to the iterator-based approach you see in `manual_serialiation.rs` to avoid O(n^2) performance when deserializing component updates. Do you know if `Schema_IsComponentUpdateFieldCleared` is guaranteed to be a constant-time lookup? If it's just walking through the list under the hood, then it's not much of an improvement, but if it's actually doing an array lookup (or something similarly efficient) then we should absolutely use it!

Reply:
Yep guaranteed to be constant time.

Reply:
I've added a `field_cleared` method to `ComponentUpdate`. Right now it's doing a linear search in the cleared fields list, but we can easily switch to using `Schema_IsComponentUpdateFieldCleared` once we update the SDK. That nicely simplified the code we need to generate for field updates 🎉